server/routes/api/schema.ts

```diff
@@ -14,7 +15,7 @@ export const ProsemirrorSchema = z.custom<TProsemirrorData>((val) => {
   try {
     const node = Node.fromJSON(schema, val);
     node.check();
-    return true;
+    return ProsemirrorHelper.isEmptyData(val as TProsemirrorData, schema);
```

Comment:
Because you're running `Node.fromJSON` inside of `isEmptyData` this is still parsing the document twice – once above on L16 and then again.

Seeing as we're already parsing to be able to validate we can use the `node` value here and pass it into `isEmpty`, something like:

```suggestion
    return !ProsemirrorHelper.isEmpty(node);
```